File: server/gemini.py
import google.generativeai as genai
import os
import json
from typing import Dict, Any, Optional

class GeminiService:
    def __init__(self):
        api_key = os.getenv("GEMINI_API_KEY") or os.getenv("GOOGLE_API_KEY")
        if not api_key:
            raise ValueError("GEMINI_API_KEY or GOOGLE_API_KEY environment variable is required")
        
        genai.configure(api_key=api_key)
        self.model = genai.GenerativeModel('gemini-pro')

    async def generate_sql_query(self, natural_language_query: str, schema: Dict[str, Any], sector: str) -> Optional[str]:
        """Generate SQL query from natural language using Gemini AI"""
        try:
            # Build schema context
            schema_context = self._build_schema_context(schema)
            
            # Build sector-specific context
            sector_context = self._build_sector_context(sector)
            
            # Create the prompt
            prompt = f"""
You are a SQL expert. Given the database schema and natural language query, generate a safe PostgreSQL query.

DATABASE SCHEMA:
{schema_context}

SECTOR CONTEXT:
{sector_context}

NATURAL LANGUAGE QUERY: {natural_language_query}

IMPORTANT RULES:
1. Generate ONLY the SQL query, no explanations
2. Use proper PostgreSQL syntax
3. Include appropriate WHERE clauses and JOINs
4. Use LIMIT to prevent returning too many rows (max 100)
5. Never use DROP, DELETE, INSERT, UPDATE, or other destructive operations
6. Only use SELECT statements
7. Handle NULL values appropriately
8. Use proper column aliases for readability

SQL Query:
"""

            response = self.model.generate_content(prompt)
            sql_query = response.text.strip()
            
            # Clean up the response
            sql_query = self._clean_sql_query(sql_query)
            
            # Validate the query is safe
            if self._is_safe_query(sql_query):
                return sql_query
            else:
                print(f"Unsafe query generated: {sql_query}")
                return None
                
        except Exception as e:
            print(f"Gemini API error: {e}")
            return None

    def _build_schema_context(self, schema: Dict[str, Any]) -> str:
        """Build schema context string for the prompt"""
        if not schema or "tables" not in schema:
            return "No schema available"
        
        context = []
        for table_name, table_info in schema["tables"].items():
            columns = table_info.get("columns", {})
            column_details = []
            
            for col_name, col_info in columns.items():
                col_type = col_info.get("type", "unknown")
                nullable = "NULL" if col_info.get("nullable") else "NOT NULL"
                column_details.append(f"  {col_name} ({col_type}) {nullable}")
            
            table_context = f"Table: {table_name}\n" + "\n".join(column_details)
            context.append(table_context)
        
        return "\n\n".join(context)

    def _build_sector_context(self, sector: str) -> str:
        """Build sector-specific context for better query generation"""
        sector_contexts = {
            "bank": """
This is a banking sector database. Common KPIs include:
- Account balances and transactions
- Loan approval rates and amounts  
- Customer demographics and banking products
- Branch performance metrics
- Risk assessments and credit scores
""",
            "finance": """
This is a financial services database. Common KPIs include:
- Revenue and profit margins
- Investment portfolio performance
- Client asset values and allocations
- Market trends and analysis
- Fee income and expense ratios
""",
            "ithr": """
This is an IT/HR database. Common KPIs include:
- Employee headcount and turnover
- Hiring metrics and time-to-fill
- Salary and compensation analysis
- Performance ratings and reviews
- Training and development metrics
"""
        }
        
        return sector_contexts.get(sector, "General business database")

    def _clean_sql_query(self, query: str) -> str:
        """Clean and format the SQL query"""
        # Remove code block markers if present
        query = query.replace("```sql", "").replace("```", "").strip()
        
        # Remove any extra whitespace
        lines = [line.strip() for line in query.split('\n') if line.strip()]
        query = ' '.join(lines)
        
        # Ensure it ends with semicolon
        if not query.endswith(';'):
            query += ';'
        
        return query

    def _is_safe_query(self, query: str) -> bool:
        """Check if the generated query is safe to execute"""
        query_upper = query.upper()
        
        # List of dangerous SQL keywords
        dangerous_keywords = [
            'DROP', 'DELETE', 'INSERT', 'UPDATE', 'ALTER', 'CREATE',
            'TRUNCATE', 'REPLACE', 'EXEC', 'EXECUTE', 'CALL'
        ]
        
        # Check for dangerous keywords
        for keyword in dangerous_keywords:
            if keyword in query_upper:
                return False
        
        # Must start with SELECT
        if not query_upper.strip().startswith('SELECT'):
            return False
        
        return True

File: server/services/gemini.ts
import { GoogleGenAI } from "@google/genai";

const ai = new GoogleGenAI({ apiKey: process.env.GOOGLE_API_KEY || "" });

export interface KPISuggestion {
  id: string;
  name: string;
  description: string;
  query_template: string;
  category: string;
}

export async function generateKPISuggestions(
  schema: any,
  sector: string
): Promise<KPISuggestion[]> {
  try {
    const schemaText = JSON.stringify(schema, null, 2);
    
    const systemPrompt = `You are an expert data analyst who generates KPI (Key Performance Indicator) suggestions based on database schemas.

Given a database schema and business sector, analyze the tables, columns, and relationships to suggest relevant KPIs.

Rules:
1. Generate 6-8 practical KPI suggestions
2. Focus on measurable business metrics relevant to the sector
3. Consider common patterns like totals, counts, averages, trends, and ratios
4. Include natural language query templates that users can ask
5. Categorize KPIs (e.g., Financial, Operational, Customer, Performance)
6. Make suggestions specific to the actual data structure available

Respond with JSON array in this exact format:
[
  {
    "id": "unique_kpi_id",
    "name": "Human-readable KPI Name",
    "description": "What this KPI measures and why it's useful",
    "query_template": "Natural language question a user would ask",
    "category": "Category name"
  }
]`;

    const prompt = `Business Sector: ${sector}

Database Schema:
${schemaText}

Based on this database schema for a ${sector} business, generate relevant KPI suggestions that can be calculated from the available data.`;

    const response = await ai.models.generateContent({
      model: "gemini-2.5-pro",
      config: {
        systemInstruction: systemPrompt,
        responseMimeType: "application/json",
      },
      contents: prompt,
    });

    const rawJson = response.text;
    console.log(`Generated KPI suggestions: ${rawJson}`);

    if (rawJson) {
      const suggestions: KPISuggestion[] = JSON.parse(rawJson);
      
      // Validate and clean the suggestions
      return suggestions.filter(suggestion => 
        suggestion.id && 
        suggestion.name && 
        suggestion.description && 
        suggestion.query_template &&
        suggestion.category
      ).slice(0, 8); // Limit to 8 suggestions
    }

    return getFallbackSuggestions(sector);
  } catch (error) {
    console.error("Failed to generate KPI suggestions with AI:", error);
    return getFallbackSuggestions(sector);
  }
}

function getFallbackSuggestions(sector: string): KPISuggestion[] {
  const fallbackSuggestions: Record<string, KPISuggestion[]> = {
    bank: [
      {
        id: "loan_approval_rate",
        name: "Loan Approval Rate",
        description: "Percentage of approved loans versus total applications",
        query_template: "What is the loan approval rate for the last 6 months?",
        category: "Operational"
      },
      {
        id: "account_growth",
        name: "Account Growth",
        description: "Rate of new account openings over time",
        query_template: "Show me the trend in new account openings",
        category: "Growth"
      },
      {
        id: "transaction_volume",
        name: "Transaction Volume",
        description: "Total transaction amounts processed",
        query_template: "What is the daily transaction volume?",
        category: "Operational"
      }
    ],
    finance: [
      {
        id: "revenue_growth",
        name: "Revenue Growth",
        description: "Revenue growth rate over time periods",
        query_template: "Show me quarterly revenue growth",
        category: "Financial"
      },
      {
        id: "profit_margins",
        name: "Profit Margins",
        description: "Profit margin analysis by product or service",
        query_template: "What are the profit margins by product category?",
        category: "Financial"
      },
      {
        id: "client_portfolio_value",
        name: "Client Portfolio Value",
        description: "Total value of client portfolios",
        query_template: "Show me client portfolio value distributions",
        category: "Financial"
      }
    ],
    ithr: [
      {
        id: "employee_turnover",
        name: "Employee Turnover Rate",
        description: "Employee turnover rate by department",
        query_template: "What is the employee turnover rate by department?",
        category: "HR Metrics"
      },
      {
        id: "hiring_metrics",
        name: "Hiring Efficiency",
        description: "Time to hire and hiring success rates",
        query_template: "Show me average time to hire by position level",
        category: "HR Metrics"
      },
      {
        id: "performance_ratings",
        name: "Performance Ratings",
        description: "Employee performance rating distributions",
        query_template: "What are the performance rating trends?",
        category: "Performance"
      }
    ]
  };

  return fallbackSuggestions[sector] || fallbackSuggestions.finance;
}

export async function generateSQLFromQuery(
  naturalLanguageQuery: string,
  schema: any,
  sector: string
): Promise<string> {
  try {
    const schemaText = JSON.stringify(schema, null, 2);
    
    const systemPrompt = `You are an expert SQL generator for ${sector} business data analysis.

Rules:
1. Generate safe, read-only SELECT queries only
2. Use proper PostgreSQL syntax
3. Include appropriate WHERE clauses for filtering
4. Use aggregate functions when appropriate (COUNT, SUM, AVG, etc.)
5. Add ORDER BY and LIMIT when helpful
6. Consider the business context of ${sector}
7. Return ONLY the SQL query with no markdown formatting, no explanations, no code blocks
8. Start the response directly with "SELECT" keyword

Available schema:
${schemaText}`;

    const response = await ai.models.generateContent({
      model: "gemini-2.5-flash",
      config: {
        systemInstruction: systemPrompt,
      },
      contents: `Generate SQL for: ${naturalLanguageQuery}

Remember: Return ONLY the SQL query starting with SELECT, no code blocks or explanations.`,
    });

    const rawSQL = response.text || "";
    console.log('Generated raw SQL:', rawSQL);
    
    // Clean up the response to ensure it's just SQL
    let cleanSQL = rawSQL.trim();
    
    // Remove any markdown formatting
    if (cleanSQL.includes('```')) {
      const sqlMatch = cleanSQL.match(/```(?:sql)?\s*(SELECT[\s\S]*?)\s*```/i);
      if (sqlMatch) {
        cleanSQL = sqlMatch[1].trim();
      }
    }
    
    // Remove any leading/trailing whitespace and ensure it starts with SELECT
    cleanSQL = cleanSQL.replace(/^[^S]*SELECT/i, 'SELECT');
    
    console.log('Cleaned SQL:', cleanSQL);
    return cleanSQL;
  } catch (error) {
    console.error("Failed to generate SQL from query:", error);
    
    // Fallback to basic SQL generation based on common patterns
    return generateFallbackSQL(naturalLanguageQuery, schema);
  }
}

function generateFallbackSQL(query: string, schema: any): string {
  const lowerQuery = query.toLowerCase();
  const tables = Object.keys(schema.tables || {});
  
  if (tables.length === 0) {
    throw new Error("No tables available for querying");
  }

  // Try to match common query patterns
  if (lowerQuery.includes("total") && lowerQuery.includes("loan")) {
    if (tables.includes("loans")) {
      return "SELECT SUM(loan_amount) as total_loan_amount, COUNT(*) as total_loans FROM loans;";
    }
  }
  
  if (lowerQuery.includes("customer") && lowerQuery.includes("count")) {
    if (tables.includes("customers")) {
      return "SELECT COUNT(*) as customer_count FROM customers;";
    }
  }
  
  if (lowerQuery.includes("payment") || lowerQuery.includes("transaction")) {
    if (tables.includes("payments")) {
      return "SELECT SUM(amount_paid) as total_payments, COUNT(*) as payment_count FROM payments;";
    }
  }
  
  if (lowerQuery.includes("loan") && lowerQuery.includes("type")) {
    if (tables.includes("loans")) {
      return "SELECT loan_type, COUNT(*) as loan_count, AVG(loan_amount) as avg_amount FROM loans GROUP BY loan_type ORDER BY loan_count DESC;";
    }
  }

  // Default fallback - show data from the first available table
  const firstTable = tables[0];
  const columns = Object.keys(schema.tables[firstTable].columns || {});
  const limitedColumns = columns.slice(0, 5).join(", "); // Limit to first 5 columns
  
  return `SELECT ${limitedColumns} FROM ${firstTable} LIMIT 10;`;
}